### PR TITLE
Log and show task history in the UI

### PR DESCRIPTION
**Pull request by clarkwinkelmann**
Work in progress

Depends on #21

### CODE REVIEW

File: src/Jobs/BasePackageJob.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+use Carbon\Carbon;
+use Flagrow\Bazaar\Composer\ComposerCommand;
+use Flagrow\Bazaar\Composer\ComposerEnvironment;
+use Flagrow\Bazaar\Composer\ComposerOutput;
+use Flagrow\Bazaar\Task;
+//use Illuminate\Bus\Queueable;
+//use Illuminate\Queue\SerializesModels;
+//use Illuminate\Queue\InteractsWithQueue;
+//use Illuminate\Contracts\Queue\ShouldQueue;
```
Comment (Luceos) [inline src/Jobs/BasePackageJob.php]: Can you take care of the commented imports?

File: src/Listeners/AddSatisConfiguration.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    /**
+     * @param TokenSet $event
+     */
+    public function setSatis(TokenSet $event)
+    {
+        $env = app()->make(ComposerEnvironment::class);
+        $editor = new ComposerFileEditor($env->getComposerJsonPath());
+        $host = FlagrowApi::getFlagrowHost() . '/api/satis';
+
+        // TODO: if a "repositories" key is already present in composer.json but uses an array instead of an object, the command quietly fails
```
Comment (Luceos) [inline src/Listeners/AddSatisConfiguration.php]: I think we need to add some functionality that identifies the "repositories" key and when it's an array mutate it into an object automatically. We could name keys based on urls or types, eg:

````json
    "workbench-1": { .. },
    "some-git.url": { .. }
```

We should add this functionality initially already to prevent confusion and errors.

File: src/Jobs/BasePackageJob.php (hunk 1 above)
Comment (clarkwinkelmann) [inline src/Jobs/BasePackageJob.php]: I'm not sure how to handle that. I could require `illuminate/queue` and use the Sync driver there. That's another dependency but that way we can use the queuable contract. 

Or I just remove them and leave the jobs as simple classes with a `handle` method

File: src/Listeners/AddSatisConfiguration.php (hunk 2 above)
Comment (clarkwinkelmann) [inline src/Listeners/AddSatisConfiguration.php]: This should probably be a separate PR based on #29

Comment (Luceos) [inline src/Listeners/AddSatisConfiguration.php]: So after merging, sounds fine with me. We do need to add it before tagging.

**Comment by clarkwinkelmann** [comment]
Sure

Le lun. 10 avr. 2017 à 12:20, Daniël Klabbers <notifications@github.com> a
écrit :

File: src/Jobs/BasePackageJob.php (hunk 1 above)
Comment (Luceos) [inline src/Jobs/BasePackageJob.php]: Queues is not yet implemented in Flarum. So calling on it would simply do nothing or trigger an error?

Comment (clarkwinkelmann) [inline src/Jobs/BasePackageJob.php]: The `queue` package of Laravel is really easy to use outside of Laravel. We could register our own queue driver (sync is the most basic one and it will work anywhere) and run the jobs through it. When Flarum implements queues we can simply switch the dispatch command to push the job to Flarum queue driver. 

At the moment I simply created a method that runs the `handle` method of the job. 

If no work has been made on queues for Flarum, maybe we should create a `flagrow/queue` package that register a queue worker and allow users to choose a driver. Then require this extension in all our extensions requiring queues.

File: src/Listeners/AddSatisConfiguration.php (hunk 2 above)
Comment (clarkwinkelmann) [inline src/Listeners/AddSatisConfiguration.php]: The "sure" answer was actually supposed to end up here. Damn you email responses

File: src/Jobs/BasePackageJob.php (hunk 1 above)
Comment (Luceos) [inline src/Jobs/BasePackageJob.php]: I've actually started a PR against core to support queues, but there was too much confusion in the team to finish it up. I wouldn't add our own (yet), unless we need it asap.

**Comment by Luceos** [comment]
How much of this is still open?

**Comment by clarkwinkelmann** [comment]
I still want to add a way for the user to clear history item and make that listing a bit prettier. We can't ship such a UI... 

Also maybe someone could check my table structure, just so I'm sure it's all right 😇

**Comment by Luceos** [comment]
Task model needs a property dates with `executed_at` in it so that it will be auto-cast to Carbon.

I've noticed a new way of firing command, but a not-so DRY way of adding tasks. Wouldn't a property on the command (or the class name) be sufficient to log into the task? Perhaps stick most of that logic into the BaseCommand instead..

**Comment by clarkwinkelmann** [comment]
@Luceos there's already a `started_at` and `finished_at` attributes. What would `executed_at` represent ?

**Comment by Luceos** [comment]
Okay, that's fine. But both need to be in `$casts` in the model for auto casting to Carbon.

**Comment by clarkwinkelmann** [comment]
Note to self: there are some changes I did not commit in `BazaarPage.js` that still ended up in the dist `extension.js` :see_no_evil: I'll push another commit with a clean dist file

**Comment by clarkwinkelmann** [comment]
I removed the WIP, I'm done with dev for this feature.

Some testing before merging would be a good idea I guess.
